src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -335,7 +335,17 @@ protected void OnDocumentContextUpdated(DocumentId documentId)
 
             if (container != null)
             {
-                OnDocumentContextUpdated(documentId, container);
+                lock (_isProjectUnloadingLock)
```

Comment:
This is now a deadlock. If OnProjectRemoved takes the serialization lock but hadn't yet taken _isProjectUnloadingLock, and then this acquires this lock here, we will call OnDocumentContextUpdated and try to take the serialization lock. The first thread now can't acquire _isProjectUnloadingLock and we're toasted.